File: May-LeetCoding-Challange-2020/Week2/997-find_the_town_judge.cpp
class Solution {
public:

    /* Time Complexity: O(n + N + N) = O(n + N)
     * Space Complexity: O(n*N) = O(n*N)
     * Run code: 0ms [[1,2]]
     * Runtime: 396 ms above 15.74%
     * Memory Usage: 66.8 MB above 12.50%
    */
   
    int findJudge(int N, vector<vector<int>>& trust) {
        
        if(N == 1)
            return 1;
        
        map<int, vector<int>> trust_map;
        int n = trust.size();
        int judge = -1;
        
        [1] -> [3]
        [1] -> [5]
        [1] - [3,5]

        for(int i = 0; i < n; i++){
        
            trust_map[trust[i][0]].push_back(trust[i][1]);
        }
        

        for(int j = 1; j <= N; j++){
            
            if(!trust_map[j].size()){
                judge = j;
            }
        }

        if(judge == -1)
            return judge;
               
        for(int k = 1; k <= N; k++){
            
            if(k != judge){
                if(!(count(trust_map[k].begin(), trust_map[k].end(), judge))){
                    return -1;
                }
            }
        }
        
        return judge;
        
    }
};

File: May-LeetCoding-Challange-2020/Week2/997-find-the-town-judge.py
class Solution(object):
    # Sol 1. Array
    def findJudge(self, N, trust):
        """
        :type N: int
        :type trust: List[List[int]]
        :rtype: int
        """
        # Base Case
        if N == 1:
            return 1

        trust_score = [0]*N
        for t in trust:
            trust_score[t[0]-1] -= 1
            trust_score[t[1]-1] += 1
      
        for i in range(len(trust_score)):
            if trust_score[i] == N-1:
                return i+1
        return -1
    
        # Time Complexity: O(t+n)
        # Space Complexity: O(n), [[1,2]]
        # Run Code Runtime: 24 ms
        # Submit Runtime: 952 ms, 16.10% faster.
        # Submit Memory Usage: 17.7 mb, 33.33% faster.